File: Assets/_Scripts/TelegraphState.cs
using System;
using System.Collections;
using UnityEngine;

public class TelegraphState : EnemyState
{
    private static readonly WaitForSeconds PopUpDelay = new WaitForSeconds(0.1f);
    private float timer;
    private Coroutine popUpCoroutine;

    public TelegraphState(Enemy enemy, EnemyStateMachine stateMachine) : base(enemy, stateMachine) { }

    public override void Enter()
    {
        timer = enemy.windupTime;
        enemy.rb.velocity = new Vector2(0, enemy.rb.velocity.y);
        if(enemy.player != null && enemy.alertSign != null)
        {
            enemy.alertSign.SetActive(true);
            enemy.alertSign.transform.localScale = Vector3.zero;
            if (popUpCoroutine != null) enemy.StopCoroutine(popUpCoroutine);
            popUpCoroutine = enemy.StartCoroutine(PopUp(enemy.alertSign));
        }
    }
    protected IEnumerator PopUp(GameObject alertSign)
    {
        float popUpSpeed = 5f;

        yield return PopUpDelay;

        while (alertSign != null && alertSign.transform.localScale.x < 1)
        {
            alertSign.transform.localScale += Vector3.one * popUpSpeed * Time.deltaTime;
            yield return null;
        }
    }


    public override void LogicUpdate()
    {
        timer -= Time.deltaTime;
        
        enemy.UpdateVisuals();

        if (timer <= 0)
        {
            stateMachine.ChangeState(enemy.attackState);
        }
    }

    public override void Exit()
    {
        if (popUpCoroutine != null)
        {
            enemy.StopCoroutine(popUpCoroutine);
            popUpCoroutine = null;
        }

        if(enemy.player != null && enemy.alertSign != null)
        {
            enemy.alertSign.SetActive(false);
        }
    }
}

File: Assets/_Scripts/Lua/telegraph_state.lua
-- ============================================================
-- 预警状态 - telegraph_state.lua
-- 
-- 对外接口（必须实现）：
--   state.name = "telegraph"
--   state.Enter(enemy) / state.LogicUpdate(enemy) / state.Exit(enemy)
--
-- C# Enemy 上可用的额外字段：
--   enemy.windupTime            -- 预警持续时间（秒）
--   enemy.alertSign             -- 警示标志 GameObject（可能为 nil）
--   enemy.alertSign:SetActive(bool)  -- 显示/隐藏
--   enemy.rb.velocity           -- 减速到 0
--
-- 典型逻辑：
--   1. Enter: 显示 alertSign，清零速度
--   2. LogicUpdate: 倒计时，到时间 → 返回 "attack"
--   3. Exit: 隐藏 alertSign
-- ============================================================
local state = {}
state.name = "telegraph"
local timer = 0

function state.Enter(enemy)
    timer = enemy.windupTime
    if enemy.alertSign ~= nil then
        enemy.alertSign:SetActive(true)
    end
    enemy.rb.velocity = CS.UnityEngine.Vector2(0, enemy.rb.velocity.y)
end

function state.LogicUpdate(enemy)
    timer = timer - CS.UnityEngine.Time.deltaTime
    if timer <= 0 then
        return "attack"
    end
    return nil
end

function state.Exit(enemy)
    if enemy.alertSign ~= nil then
        enemy.alertSign:SetActive(false)
    end
end

return state
